From 0c72713d86013fcc12c03c57f522b42c375950bc Mon Sep 17 00:00:00 2001
From: GuaiYiHu <guaiyihu@foxmail.com>
Date: Thu, 15 Oct 2026 00:31:38 +0800
Subject: [PATCH] Bluetooth: initialize GD LE capabilities only on LE
 controllers

Adapt the MiOne Android13 controller capability fix to Android14, including masked event setup and vendor capability synchronization. BR/EDR initialization and response checks remain intact.

Adapted-from: 410cfe858a1321b1737b5d7ba61bb2ae35252710
Change-Id: Ie5ff9b8b372620265e22170095f07abdb3f8c945
---
 system/gd/hci/controller.cc | 187 +++++++++++++++++++-----------------
 1 file changed, 98 insertions(+), 89 deletions(-)

diff --git a/system/gd/hci/controller.cc b/system/gd/hci/controller.cc
index 6facfef4c3..e40e720a55 100644
--- a/system/gd/hci/controller.cc
+++ b/system/gd/hci/controller.cc
@@ -58,7 +58,6 @@ struct Controller::impl {
         EventCode::NUMBER_OF_COMPLETED_PACKETS, handler->BindOn(this, &Controller::impl::NumberOfCompletedPackets));
 
     set_event_mask(kDefaultEventMask);
-    write_le_host_support(Enable::ENABLED, Enable::DISABLED);
     hci_->EnqueueCommand(ReadLocalNameBuilder::Create(),
                          handler->BindOnceOn(this, &Controller::impl::read_local_name_complete_handler));
     hci_->EnqueueCommand(ReadLocalVersionInformationBuilder::Create(),
@@ -66,14 +65,6 @@ struct Controller::impl {
     hci_->EnqueueCommand(ReadLocalSupportedCommandsBuilder::Create(),
                          handler->BindOnceOn(this, &Controller::impl::read_local_supported_commands_complete_handler));
 
-    hci_->EnqueueCommand(
-        LeReadLocalSupportedFeaturesBuilder::Create(),
-        handler->BindOnceOn(this, &Controller::impl::le_read_local_supported_features_handler));
-
-    hci_->EnqueueCommand(
-        LeReadSupportedStatesBuilder::Create(),
-        handler->BindOnceOn(this, &Controller::impl::le_read_supported_states_handler));
-
     // Wait for all extended features read
     std::promise<void> features_promise;
     auto features_future = features_promise.get_future();
@@ -83,8 +74,6 @@ struct Controller::impl {
                                              std::move(features_promise)));
     features_future.wait();
 
-    le_set_event_mask(MaskLeEventMask(local_version_information_.hci_version_, kDefaultLeEventMask));
-
     hci_->EnqueueCommand(ReadBufferSizeBuilder::Create(),
                          handler->BindOnceOn(this, &Controller::impl::read_buffer_size_complete_handler));
 
@@ -94,14 +83,30 @@ struct Controller::impl {
           handler->BindOnceOn(this, &Controller::impl::set_min_encryption_key_size_handler));
     }
 
-    if (is_supported(OpCode::LE_READ_BUFFER_SIZE_V2)) {
+    if (module_.SupportsBle()) {
+      write_le_host_support(Enable::ENABLED, Enable::DISABLED);
       hci_->EnqueueCommand(
-          LeReadBufferSizeV2Builder::Create(),
-          handler->BindOnceOn(this, &Controller::impl::le_read_buffer_size_v2_handler));
-    } else {
+          LeReadLocalSupportedFeaturesBuilder::Create(),
+          handler->BindOnceOn(this, &Controller::impl::le_read_local_supported_features_handler));
+      std::promise<void> le_features_promise;
+      auto le_features_future = le_features_promise.get_future();
       hci_->EnqueueCommand(
-          LeReadBufferSizeV1Builder::Create(),
-          handler->BindOnceOn(this, &Controller::impl::le_read_buffer_size_handler));
+          LeReadSupportedStatesBuilder::Create(),
+          handler->BindOnceOn(this, &Controller::impl::le_read_supported_states_handler,
+                              std::move(le_features_promise)));
+      le_features_future.wait();
+      le_set_event_mask(MaskLeEventMask(local_version_information_.hci_version_, kDefaultLeEventMask));
+
+      if (is_supported(OpCode::LE_READ_BUFFER_SIZE_V2)) {
+        hci_->EnqueueCommand(
+            LeReadBufferSizeV2Builder::Create(),
+            handler->BindOnceOn(this, &Controller::impl::le_read_buffer_size_v2_handler));
+      } else {
+        hci_->EnqueueCommand(
+            LeReadBufferSizeV1Builder::Create(),
+            handler->BindOnceOn(this, &Controller::impl::le_read_buffer_size_handler));
+      }
+
     }
 
     if (is_supported(OpCode::READ_LOCAL_SUPPORTED_CODECS_V1)) {
@@ -110,87 +115,90 @@ struct Controller::impl {
           handler->BindOnceOn(this, &Controller::impl::read_local_supported_codecs_v1_handler));
     }
 
-    hci_->EnqueueCommand(
-        LeReadFilterAcceptListSizeBuilder::Create(),
-        handler->BindOnceOn(this, &Controller::impl::le_read_accept_list_size_handler));
-
-    if (is_supported(OpCode::LE_READ_RESOLVING_LIST_SIZE) && module_.SupportsBlePrivacy()) {
-      hci_->EnqueueCommand(
-          LeReadResolvingListSizeBuilder::Create(),
-          handler->BindOnceOn(this, &Controller::impl::le_read_resolving_list_size_handler));
-    } else {
-      LOG_INFO("LE_READ_RESOLVING_LIST_SIZE not supported, defaulting to 0");
-      le_resolving_list_size_ = 0;
-    }
-
-    if (is_supported(OpCode::LE_READ_MAXIMUM_DATA_LENGTH) && module_.SupportsBleDataPacketLengthExtension()) {
-      hci_->EnqueueCommand(LeReadMaximumDataLengthBuilder::Create(),
-                           handler->BindOnceOn(this, &Controller::impl::le_read_maximum_data_length_handler));
-    } else {
-      LOG_INFO("LE_READ_MAXIMUM_DATA_LENGTH not supported, defaulting to 0");
-      le_maximum_data_length_.supported_max_rx_octets_ = 0;
-      le_maximum_data_length_.supported_max_rx_time_ = 0;
-      le_maximum_data_length_.supported_max_tx_octets_ = 0;
-      le_maximum_data_length_.supported_max_tx_time_ = 0;
-    }
-
-    // SSP is managed by security layer once enabled
-    write_simple_pairing_mode(Enable::ENABLED);
     if (module_.SupportsSecureConnections()) {
       hci_->EnqueueCommand(
           WriteSecureConnectionsHostSupportBuilder::Create(Enable::ENABLED),
           handler->BindOnceOn(
               this, &Controller::impl::write_secure_connections_host_support_complete_handler));
     }
-    if (is_supported(OpCode::LE_READ_SUGGESTED_DEFAULT_DATA_LENGTH) && module_.SupportsBleDataPacketLengthExtension()) {
+    if (module_.SupportsBle()) {
       hci_->EnqueueCommand(
-          LeReadSuggestedDefaultDataLengthBuilder::Create(),
-          handler->BindOnceOn(this, &Controller::impl::le_read_suggested_default_data_length_handler));
-    } else {
-      LOG_INFO("LE_READ_SUGGESTED_DEFAULT_DATA_LENGTH not supported, defaulting to 27 (0x1B)");
-      le_suggested_default_data_length_ = 27;
-    }
+          LeReadFilterAcceptListSizeBuilder::Create(),
+          handler->BindOnceOn(this, &Controller::impl::le_read_accept_list_size_handler));
 
-    if (is_supported(OpCode::LE_READ_MAXIMUM_ADVERTISING_DATA_LENGTH) && module_.SupportsBleExtendedAdvertising()) {
-      hci_->EnqueueCommand(
-          LeReadMaximumAdvertisingDataLengthBuilder::Create(),
-          handler->BindOnceOn(this, &Controller::impl::le_read_maximum_advertising_data_length_handler));
-    } else {
-      LOG_INFO("LE_READ_MAXIMUM_ADVERTISING_DATA_LENGTH not supported, defaulting to 31 (0x1F)");
-      le_maximum_advertising_data_length_ = 31;
-    }
+      if (is_supported(OpCode::LE_READ_RESOLVING_LIST_SIZE) && module_.SupportsBlePrivacy()) {
+        hci_->EnqueueCommand(
+            LeReadResolvingListSizeBuilder::Create(),
+            handler->BindOnceOn(this, &Controller::impl::le_read_resolving_list_size_handler));
+      } else {
+        LOG_INFO("LE_READ_RESOLVING_LIST_SIZE not supported, defaulting to 0");
+        le_resolving_list_size_ = 0;
+      }
 
-    if (is_supported(OpCode::LE_READ_NUMBER_OF_SUPPORTED_ADVERTISING_SETS) &&
-        module_.SupportsBleExtendedAdvertising()) {
-      hci_->EnqueueCommand(
-          LeReadNumberOfSupportedAdvertisingSetsBuilder::Create(),
-          handler->BindOnceOn(this, &Controller::impl::le_read_number_of_supported_advertising_sets_handler));
-    } else {
-      LOG_INFO("LE_READ_NUMBER_OF_SUPPORTED_ADVERTISING_SETS not supported, defaulting to 1");
-      le_number_supported_advertising_sets_ = 1;
-    }
+      if (is_supported(OpCode::LE_READ_MAXIMUM_DATA_LENGTH) && module_.SupportsBleDataPacketLengthExtension()) {
+        hci_->EnqueueCommand(LeReadMaximumDataLengthBuilder::Create(),
+                             handler->BindOnceOn(this, &Controller::impl::le_read_maximum_data_length_handler));
+      } else {
+        LOG_INFO("LE_READ_MAXIMUM_DATA_LENGTH not supported, defaulting to 0");
+        le_maximum_data_length_.supported_max_rx_octets_ = 0;
+        le_maximum_data_length_.supported_max_rx_time_ = 0;
+        le_maximum_data_length_.supported_max_tx_octets_ = 0;
+        le_maximum_data_length_.supported_max_tx_time_ = 0;
+      }
 
-    if (is_supported(OpCode::LE_READ_PERIODIC_ADVERTISER_LIST_SIZE) &&
-        module_.SupportsBlePeriodicAdvertising()) {
-      hci_->EnqueueCommand(
-          LeReadPeriodicAdvertiserListSizeBuilder::Create(),
-          handler->BindOnceOn(this, &Controller::impl::le_read_periodic_advertiser_list_size_handler));
-    } else {
-      LOG_INFO("LE_READ_PERIODIC_ADVERTISER_LIST_SIZE not supported, defaulting to 0");
-      le_periodic_advertiser_list_size_ = 0;
-    }
-    if (is_supported(OpCode::LE_SET_HOST_FEATURE) && module_.SupportsBleConnectedIsochronousStreamCentral()) {
-      hci_->EnqueueCommand(
-          LeSetHostFeatureBuilder::Create(LeHostFeatureBits::CONNECTED_ISO_STREAM_HOST_SUPPORT, Enable::ENABLED),
-          handler->BindOnceOn(this, &Controller::impl::le_set_host_feature_handler));
-    }
+      // SSP is managed by security layer once enabled
+      write_simple_pairing_mode(Enable::ENABLED);
+      if (is_supported(OpCode::LE_READ_SUGGESTED_DEFAULT_DATA_LENGTH) && module_.SupportsBleDataPacketLengthExtension()) {
+        hci_->EnqueueCommand(
+            LeReadSuggestedDefaultDataLengthBuilder::Create(),
+            handler->BindOnceOn(this, &Controller::impl::le_read_suggested_default_data_length_handler));
+      } else {
+        LOG_INFO("LE_READ_SUGGESTED_DEFAULT_DATA_LENGTH not supported, defaulting to 27 (0x1B)");
+        le_suggested_default_data_length_ = 27;
+      }
+
+      if (is_supported(OpCode::LE_READ_MAXIMUM_ADVERTISING_DATA_LENGTH) && module_.SupportsBleExtendedAdvertising()) {
+        hci_->EnqueueCommand(
+            LeReadMaximumAdvertisingDataLengthBuilder::Create(),
+            handler->BindOnceOn(this, &Controller::impl::le_read_maximum_advertising_data_length_handler));
+      } else {
+        LOG_INFO("LE_READ_MAXIMUM_ADVERTISING_DATA_LENGTH not supported, defaulting to 31 (0x1F)");
+        le_maximum_advertising_data_length_ = 31;
+      }
+
+      if (is_supported(OpCode::LE_READ_NUMBER_OF_SUPPORTED_ADVERTISING_SETS) &&
+          module_.SupportsBleExtendedAdvertising()) {
+        hci_->EnqueueCommand(
+            LeReadNumberOfSupportedAdvertisingSetsBuilder::Create(),
+            handler->BindOnceOn(this, &Controller::impl::le_read_number_of_supported_advertising_sets_handler));
+      } else {
+        LOG_INFO("LE_READ_NUMBER_OF_SUPPORTED_ADVERTISING_SETS not supported, defaulting to 1");
+        le_number_supported_advertising_sets_ = 1;
+      }
+
+      if (is_supported(OpCode::LE_READ_PERIODIC_ADVERTISER_LIST_SIZE) &&
+          module_.SupportsBlePeriodicAdvertising()) {
+        hci_->EnqueueCommand(
+            LeReadPeriodicAdvertiserListSizeBuilder::Create(),
+            handler->BindOnceOn(this, &Controller::impl::le_read_periodic_advertiser_list_size_handler));
+      } else {
+        LOG_INFO("LE_READ_PERIODIC_ADVERTISER_LIST_SIZE not supported, defaulting to 0");
+        le_periodic_advertiser_list_size_ = 0;
+      }
+      if (is_supported(OpCode::LE_SET_HOST_FEATURE) && module_.SupportsBleConnectedIsochronousStreamCentral()) {
+        hci_->EnqueueCommand(
+            LeSetHostFeatureBuilder::Create(LeHostFeatureBits::CONNECTED_ISO_STREAM_HOST_SUPPORT, Enable::ENABLED),
+            handler->BindOnceOn(this, &Controller::impl::le_set_host_feature_handler));
+      }
+
+      if (common::init_flags::subrating_is_enabled() && is_supported(OpCode::LE_SET_HOST_FEATURE) &&
+          module_.SupportsBleConnectionSubrating()) {
+        hci_->EnqueueCommand(
+            LeSetHostFeatureBuilder::Create(
+                LeHostFeatureBits::CONNECTION_SUBRATING_HOST_SUPPORT, Enable::ENABLED),
+            handler->BindOnceOn(this, &Controller::impl::le_set_host_feature_handler));
+      }
 
-    if (common::init_flags::subrating_is_enabled() && is_supported(OpCode::LE_SET_HOST_FEATURE) &&
-        module_.SupportsBleConnectionSubrating()) {
-      hci_->EnqueueCommand(
-          LeSetHostFeatureBuilder::Create(
-              LeHostFeatureBits::CONNECTION_SUBRATING_HOST_SUPPORT, Enable::ENABLED),
-          handler->BindOnceOn(this, &Controller::impl::le_set_host_feature_handler));
     }
 
     if (is_supported(OpCode::READ_DEFAULT_ERRONEOUS_DATA_REPORTING)) {
@@ -201,7 +209,7 @@ struct Controller::impl {
     }
 
     // Skip vendor capabilities check if configured.
-    if (os::GetSystemPropertyBool(
+    if (module_.SupportsBle() && os::GetSystemPropertyBool(
             kPropertyVendorCapabilitiesEnabled, kDefaultVendorCapabilitiesEnabled)) {
       // More commands can be enqueued from le_get_vendor_capabilities_handler
       std::promise<void> vendor_promise;
@@ -490,12 +498,13 @@ struct Controller::impl {
     le_local_supported_features_ = complete_view.GetLeFeatures();
   }
 
-  void le_read_supported_states_handler(CommandCompleteView view) {
+  void le_read_supported_states_handler(std::promise<void> promise, CommandCompleteView view) {
     auto complete_view = LeReadSupportedStatesCompleteView::Create(view);
     ASSERT(complete_view.IsValid());
     ErrorCode status = complete_view.GetStatus();
     ASSERT_LOG(status == ErrorCode::SUCCESS, "Status 0x%02hhx, %s", status, ErrorCodeText(status).c_str());
     le_supported_states_ = complete_view.GetLeStates();
+    promise.set_value();
   }
 
   void le_read_accept_list_size_handler(CommandCompleteView view) {
